02-peer-protocol.md

```diff
@@ -50,10 +67,365 @@ pubkey corresponding to the funding output nothing prevents duplicative channel
+received `serial_id` before forwarding them to other peers.
+
+Here are a few example exchanges.
+
+#### *initiator* only
+
+A, *initiator* has two inputs and an output (the funding output).
+B, the *non-initiator* has nothing to contribute.
+
+        +-------+                       +-------+
+        |       |--(1)- tx_add_input -->|       |
+        |       |<-(2)- tx_complete ----|       |
+        |       |--(3)- tx_add_input -->|       |
+        |   A   |<-(4)- tx_complete ----|   B   |
+        |       |--(5)- tx_add_output ->|       |
+        |       |                       |       |
+        |       |--(6)- tx_complete --->|       |
+        |       |<-(7)- tx_complete ----|       |
+        +-------+                       +-------+
+
+#### *initiator* and *non-initiator*
+
+A the *initiator* contributes 2 inputs and an output that they
+then remove.  B, the *non-initiator*, contributes 1 input and an output,
+but waits until A adds a second input before contributing.
+
+Note that if A does not send a second input, the negotiation will end without
+B's contributions.
+
+        +-------+                         +-------+
+        |       |--(1)- tx_add_input ---->|       |
+        |       |<-(2)- tx_complete ------|       |
+        |       |--(3)- tx_add_output --->|       |
+        |       |<-(4)- tx_complete ------|       |
+        |       |--(5)- tx_add_input ---->|       |
+        |   A   |<-(6)- tx_add_input -----|   B   |
+        |       |<-(7)- tx_add_output ----|       |
+        |       |--(8)- tx_remove_output >|       |
+        |       |<-(9)- tx_complete ------|       |
+        |       |--(10) tx_complete ----->|       |
+        +-------+                         +-------+
+
+
+### The `tx_add_input` Message
+
+This message contains a transaction input.
+
+1. type: 66 (`tx_add_input`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+    * [`u16`:`prevtx_len`]
+    * [`prevtx_len*byte`:`prevtx`]
+    * [`u32`:`prevtx_vout`]
+    * [`u32`:`sequence`]
+    * [`u16`:`script_sig_len`]
+    * [`script_sig_len*byte`:`script_sig`]
+
+#### Requirements
+
+The sending node:
+  - MUST add all sent inputs to the transaction
+  - MUST use a unique `serial_id` for each input currently added to the
+    transaction
+  - MUST NOT re-transmit inputs it has received from the peer
+  - if is the *initiator*:
+    - MUST send even `serial_id`s
+  - if is the *non-initiator*:
+    - MUST send odd `serial_id`s
+
+The receiving node:
+  - MUST add all received inputs to the transaction
+  - MUST fail the negotiation if:
+    - the `prevtx` and `prevtx_vout` are identical to a previously added
+      (and not removed) input's
+    - `prevtx` is not a valid transaction
+    - `prevtx_vout` is greater or equal to the number of outputs on `prevtx`
+    - the `scriptPubKey` of the `prevtx_out` output of `prevtx` does not begin with `OP_0` to `OP_16`
+      followed by a single push of 2 to 40 bytes
```

Comment:
What is the rationale for not allowing p2sh-embedded segwit inputs?

Reply:
We should remove support for non-segwit inputs, as it creates a lot of issues when allowing dust limits below 546 sats (see https://github.com/lightningnetwork/lightning-rfc/pull/894)

Reply:
Oh I see. This should be updated to be *either* the `scriptPubKey` or the `script_sig`, if provided. This will encapsulate both cases, and permit P2SH-wrapped segwit spends to be eligible.

Reply:
Ok updated the text to allow for P2SH-wrapped spends.

Regarding moving to native segwit only (P2W*++), this is a reasonable policy for *outputs* I think, but not for spending P2SH-wrapped inputs, as there's no dust considerations for spends (unlike output creations)